### PR TITLE
[Merged by Bors] - feat(Data/Finsupp): make toMultiset and antidiagonal computable

### PR DESCRIPTION
**Pull request by eric-wieser**
In Lean 3, the computability of `Finsupp.toMultiset` was poisoned by the `AddMonoid (α →₀ ℕ)` instance, even though this was not used in computation. This is no longer the case in Lean 4, so we can make this computable by adding a `DecidableEq α` argument.

We loosely follow the pattern used with `DFinsupp`, where we split the declaration in two, as only one direction needs `DecidableEq α`. As a result, `Finsupp.toMultiset` is now only an `AddMonoidHom`, though `Multiset.toFinset` remains an equiv.

We're missing some of the formatting infrastructure for this to be pretty, but this now works:
```lean
#eval ((Finsupp.mk Finset.univ ![1, 2, 3] sorry).antidiagonal).image
  fun x : _ × _ => (x.1.toFun, x.2.toFun)
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by ericrbg — APPROVED** [review]
lgtm, I like the way to avoid a naming clash whilst still having short names

**Comment by ericrbg** [comment]
maintainer merge

File: Mathlib/RingTheory/PowerSeries/Basic.lean [hunk 1]
```diff
@@ -1473,6 +1474,7 @@ theorem coeff_zero_one : coeff R 0 (1 : PowerSeries R) = 1 :=
 theorem coeff_mul (n : ℕ) (φ ψ : PowerSeries R) :
     coeff R n (φ * ψ) = ∑ p in Finset.Nat.antidiagonal n, coeff R p.1 φ * coeff R p.2 ψ := by
   symm
+  erw [MvPowerSeries.coeff_mul]
```
Comment (Vierkantor) [inline Mathlib/RingTheory/PowerSeries/Basic.lean:1477]: Could you explain (in a comment?) why a normal `rw` doesn't work here?

**Review by Vierkantor — APPROVED** [review]
Nice work!

bors d+

File: Mathlib/RingTheory/PowerSeries/Basic.lean [hunk 2]
```diff
@@ -1473,6 +1474,7 @@ theorem coeff_zero_one : coeff R 0 (1 : PowerSeries R) = 1 :=
 theorem coeff_mul (n : ℕ) (φ ψ : PowerSeries R) :
     coeff R n (φ * ψ) = ∑ p in Finset.Nat.antidiagonal n, coeff R p.1 φ * coeff R p.2 ψ := by
   symm
+  erw [MvPowerSeries.coeff_mul] -- `rw` can't see that `PowerSeries = MvPowerSeries Unit`
```
Comment (ericrbg) [inline Mathlib/RingTheory/PowerSeries/Basic.lean:1477]: not to be a nitpicker but in this case shouldn't we just have a lemma to this effect and normal `rw` it? [or does that cause motive issues]

the defeq abuse before wasn't great either though

Comment (eric-wieser) [reply Mathlib/RingTheory/PowerSeries/Basic.lean]: This lemma is the place where we're building that lemma! The problem would go away if we made PowerSeries reducible, or introduced a function to convert between it and MvPowerSeries

**Comment by eric-wieser** [comment]
bors merge

File: Mathlib/RingTheory/PowerSeries/Basic.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/RingTheory/PowerSeries/Basic.lean]: I'll make a follow-up PR to golf this proof

**Pull request closed without merging**